app/javascript/components/common/svg-graph-util.ts

```diff
@@ -7,119 +7,221 @@ export type Line = {
   length: number
   angle: number
 }
-
-const SMOOTHING_RATIO = 0.2
+const SIMILAR_THRESHOLD_DELTA = 0.01
+const SMOOTHING_THRESHOLD_DELTA = 0.2
+const INTERPOLATION_STEPS = 3
+const SEGMENT_PARTS = INTERPOLATION_STEPS + 1
 
 export const minMaxNormalize = (data: Array<number>): Array<number> => {
   const min = Math.min(0, ...data)
   const max = Math.max(...data)
+
+  if (max === 0 && min === 0) {
+    return data
+  }
+
   return data.map((n) => (n - min) / (max - min))
 }
 
-export const drawSegmentPath = (dataPoints: Array<DataPoint>): string => {
+export const mapToSvgDimensions = (
+  data: number[],
+  height: number,
+  width: number
+): DataPoint[] => {
+  // due to the nature of bezier splines, some curves can go outside the boundaries without this buffer
+  const vBuffer = height * 0.05
+  const step = width / (data.length - 1)
+
+  return data.map((normalizedHeight: number, index: number) => {
+    if (normalizedHeight < 0 || normalizedHeight > 1) {
+      throw new Error('Normalized height must be between 0 and 1')
+    }
+
+    return {
+      x: index * step,
+      y: height - vBuffer - (height - 2 * vBuffer) * normalizedHeight, // SVG coordinates start from the upper left corner
+    } as DataPoint
+  })
+}
+
+/**
+ * With splines (bezier or catmull-rom) when there are large differences in values, in order to
+ * keep the curve smooth, it often will vary above or below the baseline since they must touch all
+ * of the points.  This function examines the list of data points, and if there is an extreme difference
+ * attempts to prevent the dipping above or below the baseline by interpolating points to guide the curve
+ */
+export const smoothDataPoints = (
+  dataPoints: DataPoint[],
+  height: number
+): DataPoint[] => {
   return dataPoints.reduce(
     (
-      path: string,
+      acc: DataPoint[],
       dataPoint: DataPoint,
-      idx: number,
-      dataPoints: Array<DataPoint>
+      index: number,
+      dataPoints: readonly DataPoint[]
     ) => {
-      switch (idx) {
-        case 0:
-          return `${drawM(dataPoint)}`
-        case 1:
-          return `${path} ${drawC(dataPoints, idx)}`
-        default:
-          return `${path} ${drawS(dataPoints, idx)}`
+      const prevPoint = dataPoints[index - 1]
+      const nextPoint = dataPoints[index + 1]
+      const nextNextPoint = dataPoints[index + 2]
```

Comment:
if the current index is near the beginning or the end, these may be undefined, which is why they are being handled carfully on lines 69 and 77